src/encryption_layer.rs

```diff
@@ -0,0 +1,248 @@
+use std::{fmt::Debug, ops::Deref, sync::Arc};
+
+use crate::{
+    ADDRESS_LENGTH, KEY_LENGTH, MemoryADT, address::Address, error::Error, secret::Secret,
+    symmetric_key::SymmetricKey,
+};
+use aes::{
+    Aes256,
+    cipher::{BlockEncrypt, KeyInit, generic_array::GenericArray},
+};
+use xts_mode::Xts128;
+
+#[derive(Clone)]
+struct ClonableXts(Arc<Xts128<Aes256>>);
+
+impl Debug for ClonableXts {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_tuple("ClonableXts").finish()
+    }
+}
+
+impl Deref for ClonableXts {
+    type Target = Xts128<Aes256>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+/// The encryption layers is built on top of an encrypted memory implementing the `MemoryADT` and
+/// exposes a plaintext virtual memory interface implementing the `MemoryADT`.
+///
+/// This type is thread-safe.
+#[derive(Debug, Clone)]
+pub struct MemoryEncryptionLayer<
+    const WORD_LENGTH: usize,
+    Memory: MemoryADT<Address = Address<ADDRESS_LENGTH>>,
+> {
+    aes: Aes256,
+    xts: ClonableXts,
+    mem: Memory,
+}
+
+impl<
+    const WORD_LENGTH: usize,
+    Memory: Send + Sync + MemoryADT<Address = Address<ADDRESS_LENGTH>, Word = [u8; WORD_LENGTH]>,
+> MemoryEncryptionLayer<WORD_LENGTH, Memory>
+{
+    /// Instantiates a new memory encryption layer.
+    pub fn new(seed: Secret<KEY_LENGTH>, stm: Memory) -> Self {
+        let k_p = SymmetricKey::<KEY_LENGTH>::derive(&seed, &[0]).expect("secret is large enough");
```

Comment:
I suppose the notations are the ones from the article. Could you link the pdf file somewhere?